### ISSUE TITLE
nvidia pods for driver, device plugin, dcgm exporter, feature discovery in crash loop after host reboot

### ISSUE THREAD

**Issue by tusharrobin** [NVIDIA/gpu-operator#197]
_The template below is mostly useful for bug reports and support questions. Feel free to remove anything which doesn't apply to you and add more information where it makes sense._

### 1. Quick Debug Checklist
- [ ] Are you running on an Ubuntu 18.04 node? => Centos 7.9
- [x ] Are you running Kubernetes v1.13+?  => v1.20.5
- [x] Are you running Docker (>= 18.06) or CRIO (>= 1.13+)? ==>   19.03.6
- [ ] Do you have `i2c_core` and `ipmi_msghandler` loaded on the nodes?
- [x] Did you apply the CRD (`kubectl describe clusterpolicies --all-namespaces`) ==> Yes, its deployed via GPU operator

### 1. Issue or feature description
On Centos 7.9, GPU operator installed successfully and all pods became ready but after rebooting of node, pods went into crashloopbackoff state.

gpu-operator-resources   gpu-feature-discovery-qcdp4                                       0/1     Init:CrashLoopBackOff   10         44m
gpu-operator-resources   nvidia-container-toolkit-daemonset-rkg4b                          0/1     Init:CrashLoopBackOff   10         44m
gpu-operator-resources   nvidia-cuda-validator-ssgbh                                       0/1     Completed               0          42m
gpu-operator-resources   nvidia-dcgm-exporter-kj45b                                        0/1     Init:CrashLoopBackOff   10         44m
gpu-operator-resources   nvidia-device-plugin-daemonset-zdc4w                              0/1     Init:CrashLoopBackOff   11         44m
gpu-operator-resources   nvidia-device-plugin-validator-qbhtk                              0/1     Completed               0          42m
gpu-operator-resources   nvidia-driver-daemonset-svsmn                                     0/1     CrashLoopBackOff        10         44m
gpu-operator-resources   nvidia-operator-validator-j9m2z                                   0/1     Init:CrashLoopBackOff   11         44m

### 2. Steps to reproduce the issue
1. Install Kubernetes 
2. Install GPU operator and make sure all pods are running. Test sample pod with GPU to see if everything is working.
3. Reboot node 
4. Pod go into crashloopbackoff

### 3. Information to [attach](https://help.github.com/articles/file-attachments-on-issues-and-pull-requests/) (optional if deemed irrelevant)

 - [ ] kubernetes pods status: `kubectl get pods --all-namespaces`
 - [ ] kubernetes daemonset status: `kubectl get ds --all-namespaces`
 - [x] If a pod/ds is in an error state or pending state `kubectl describe pod -n NAMESPACE POD_NAME`
kubectl describe pod -n gpu-operator-resources   nvidia-driver-daemonset-svsmn

  Warning  BackOff         3m16s (x143 over 34m)  kubelet            Back-off restarting failed container

      Reason:       ContainerCannotRun
      **Message:      OCI runtime create failed: container_linux.go:349: starting container process caused "process_linux.go:449: container init caused \"process_linux.go:432: running prestart hook 0 caused \\\"error running hook: exit status 1, stdout: , stderr: nvidia-container-cli.real: initialization error: driver error: failed to process request\\\\n\\\"\"": unknown**

 - [] If a pod/ds is in an error state or pending state `kubectl logs -n NAMESPACE POD_NAME`

 - [x] Output of running a container on the GPU machine: `docker run -it alpine echo foo`
 - [ ] 
 $ docker run -it alpine echo foo
foo

 - [ ] Docker configuration file: `cat /etc/docker/daemon.json`
 - [x] Docker runtime configuration: `docker info | grep runtime`

docker info| grep -i runtime
 Runtimes: nvidia runc
 Default Runtime: nvidia

 - [ ] NVIDIA shared directory: `ls -la /run/nvidia`
 - [ ] NVIDIA packages directory: `ls -la /usr/local/nvidia/toolkit`
 - [ ] NVIDIA driver directory: `ls -la /run/nvidia/driver`
 - [ ] kubelet logs `journalctl -u kubelet > kubelet.logs`

**Comment by tusharrobin** [comment]
@shivamerla I am using toolkit.version=1.4.0-ubi8. I am using 1.4.0 since I faced glibc issues mentioned in https://github.com/NVIDIA/gpu-operator/issues/72 when using default toolkit version shipped with 1.7.

**Comment by bstrdsmkr** [comment]
I'm working through this same problem after attempting an upgrade from helm chart version 1.6.0. I had this working using the 1.4.7-ubi8 toolkit. FYI, the available toolkit versions can be found here: https://hub.docker.com/r/nvidia/container-toolkit/tags?page=1&ordering=last_updated

It looks like 1.5.0 is the most recent. Theoretically, the "-ubi8" images are what you need for CentOS, though you'll have to do a bit of trial and error to find one compiled with a matching version of GLIBC

Also, according to the previous bugs, there were plans to have 1.5.0 compiled with a statically linked GLIBC though that apparently didn't happen: https://github.com/NVIDIA/libnvidia-container/issues/81

**Comment by shivamerla** [comment]
Yes, thats the best option to switch to `1.5.0-ubi8` toolkit version. To recover from current state, you can change default-runtime to `runc` in docker daemon.json and restart docker. Toolkit will set the runtime again on startup.

**Comment by bstrdsmkr** [comment]
@shivamerla can we reopen this? I'm seeing this issue still on a RHEL 7.9 host with toolkit version 1.5.0-ubi8

**Comment by shivamerla** [comment]
@bstrdsmkr sure, i will re-open this. So, the solution works only with `1.4.7-ubi8` toolkit version?

**Comment by bstrdsmkr** [comment]
RHEL 7.8 worked with 1.4.7-ubi, but neither appears to work with RHEL 7.9 sadly :(

**Comment by shivamerla** [comment]
got it. seems like RHEL7.9 GLIBC version doesn't match with one packaged with either of those toolkit images. statically linking nvidia-container-cli will be the proper fix for this. Will try to get this fixed soon.

**Comment by bstrdsmkr** [comment]
@shivamerla thanks for that!

I'm definitely not great with makefiles, but looking at this comment: https://github.com/NVIDIA/libnvidia-container/issues/81#issuecomment-751505203 and this makefile: https://github.com/NVIDIA/libnvidia-container/blob/master/Makefile it looks to me like I could build a custom image by cloning the repo and running `make -f mk/docker.mk centos7 BIN_LDLIBS='-l:$(LIB_STATIC) -lseccomp -lcap -ldl'` or possibly `make -f mk/docker.mk centos7.9.2009` (based on the tags available for base images here: https://hub.docker.com/_/centos?tab=tags&page=1&ordering=last_updated)

Does that seem like an appropriate path or have I missed something?

**Comment by shivamerla** [comment]
I will let @klueska @elezar comment on this.

**Comment by elezar** [comment]
@bstrdsmkr unfortunately defining the contents of the make variables from the command line as you propose will not work as these are not conditionally set in the make files.

If you are building a custom image, what would be the reason for building statically? Would the `centos7-x86_64` target not be sufficient in your case or is static linking a must-have even if you are able to produce a compatible library?

**Comment by bstrdsmkr** [comment]
@elezar The problem I'm trying to solve is the toolkit crashlooping:
```
Error: failed to create containerd task: OCI runtime create failed: container_linux.go:380: starting container process caused: process_linux.go:545: contai
ner init caused: Running hook #0:: error running hook: exit status 1, stdout: , stderr: /usr/local/nvidia/toolkit/nvidia-container-cli.real: /lib64/libc.so.6: version 'GLIBC_2.27' not found (required by /usr/local
/nvidia/toolkit/libnvidia-container.so.1): unknown
```

This is a CentOS 7.9 node using either to 1.4.7-ubi8 or 1.5.0-ubi8 toolkit images. On CentOS 7.8, 1.4.7-ubi8 worked fine, but it appears that none of the ubi images are linked against CentOS 7.9's glibc 2.17:
```
[root@arm-login01 containerd]# ldd --version
ldd (GNU libc) 2.17
Copyright (C) 2012 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```
```
[root@arm-login01 containerd]# cat /etc/*release
CentOS Linux release 7.9.2009 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

CentOS Linux release 7.9.2009 (Core)
CentOS Linux release 7.9.2009 (Core)
[root@arm-login01 containerd]#
```
I take from this thread: https://github.com/NVIDIA/libnvidia-container/issues/81#issuecomment-751711356 that building a staticly linked binary is the long term fix for this problem, though until now using the ubi images were an acceptable workaround. Until there is solve for this, our ML platform is is broken :(

**Comment by janeczku** [comment]
Given that there is no workaround that would allow GPU operator to work on CentOS 7.9, is there a timeline for shipping the fix mentioned [above](https://github.com/NVIDIA/gpu-operator/issues/197#issuecomment-854100357) ?

**Comment by elezar** [comment]
Hi @janeczku there are some issues with the fix you mention and we are trying to determine what the best way forward is. We will be looking at releasing an opt-in option for an alternative implementation that should also address these issue in the case of the GPU Operator. If you don't mind creating a separate issue that describes your use case and options / features used in GPU containers, we could see if this is a good fit for the opt-in feature.

**Comment by janeczku** [comment]
Hi and thanks for feedback @elezar I am happy to describe the use case but not sure how this is relevant here:
CentOS 7 is explicitly supported by the GPU operator per the [official docs](https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/platform-support.html#linux-distributions)  . Given that deploying the operator on CentOS 7.9 fails (see OP) i would expect this to be addressed by a bug fix rather than an "opt-in" feature.

**Comment by elezar** [comment]
@janeczku that's a fair point. We will push out centos7 based images of the toolkit-container for use with the operator as soon as possible. These can then be selected on Centos7 systems by specifying a `toolkit.version` when installing the operator.

**Comment by janeczku** [comment]
Thanks for the clarification @elezar 👍

**Comment by reddiablo85** [comment]
Hi...we encountered this issue with RHEL 7.9 and subsequently updated our nodes to 8.4....same problem occuring....using the "XXXX.ubi" images helped to get the other up but the dcgm-exporter still fails....see below events for the dcgm-exporter:

Events:
  Type     Reason       Age                  From                      Message
  ----     ------       ----                 ----                      -------
  Normal   Scheduled    <unknown>                                      Successfully assigned gpu-operator-resources/nvidia-dcgm-exporter-h6dj5 to gdcslxrrawo303t
  Warning  FailedMount  2m59s                kubelet, gdcslxrrawo303t  Unable to attach or mount volumes: unmounted volumes=[init-config], unattached volumes=[run-nvidia nvidia-dcgm-exporter-token-qqsrx pod-gpu-resources init-config]: timed out waiting for the condition
  Warning  FailedMount  52s (x10 over 5m2s)  kubelet, gdcslxrrawo303t  MountVolume.SetUp failed for volume "init-config" : configmap "nvidia-dcgm-exporter" not found
  Warning  FailedMount  43s                  kubelet, gdcslxrrawo303t  Unable to attach or mount volumes: unmounted volumes=[init-config], unattached volumes=[pod-gpu-resources init-config run-nvidia nvidia-dcgm-exporter-token-qqsrx]: timed out waiting for the condition

You can see that it is trying to inject a config that doesn´t exist. I have scoured the helm files and can only see mention of this init-config for the openshift platform. We´re running in k8s. Is this perhaps missing?

**Comment by klueska** [comment]
> got it. seems like RHEL7.9 GLIBC version doesn't match with one packaged with either of those toolkit images. statically linking nvidia-container-cli will be the proper fix for this. Will try to get this fixed soon.

I evaluated statically linking this binary a few months and there are more issues with doing that than leaving it as a dynamic library abs „bundling“ a specific libc (and other dependencies). I have an internal JIRA with the details of my findings and a proposal due the „proper“ fix.

**Comment by elezar** [comment]
@janeczku we have test images for the `container-toolkit` out that you could test if you have time. The image can be specified by adding `--set toolkit.version="devel-centos7"` when installing the operator using `helm`. If all goes well we will include these images in an upcoming toolkit container release so that a "proper" version tag can be used.

**Comment by bstrdsmkr** [comment]
@elezar just gave it a spin and it looks like it's still looking for glibc 2.27 on the host:

```
  Normal   Pulled       10s (x2 over 22s)  kubelet            Container image "nvcr.io/nvidia/k8s/container-toolkit:devel-centos7" already present on machine                                                                                                                                
  Warning  Failed       9s (x3 over 23s)   kubelet            Error: failed to create containerd task: OCI runtime create failed: container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: Running hook #0:: error running hook: exit status 1
, stdout: , stderr: /usr/local/nvidia/toolkit/nvidia-container-cli.real: /lib64/libc.so.6: version `GLIBC_2.27' not found (required by /usr/local/nvidia/toolkit/libnvidia-container.so.1): unknown                                                                                          
  Warning  BackOff      9s (x2 over 21s)   kubelet            Back-off restarting failed container
```

Let me know what other info I can provide to help

**Comment by elezar** [comment]
@bstrdsmkr thanks for checking. Could it be that the `libnvidia-container.so.*` files are remnants of another deployment? I have run:
```
# objdump -T /usr/lib64/libnvidia-container.so.1.4.0 | grep GLIBC_2.27
```
in the `devel-centos7` image and confirmed that no `GLIBC_2.27` symbols are present.

For the `ubuntu18.04` image (after installing `binutils) we see:
```
objdump -T /usr/lib/x86_64-linux-gnu/libnvidia-container.so.1.4.0 | grep GLIBC_2.27
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.27  glob
```

**Comment by bstrdsmkr** [comment]
@elezar Success! You were correct, I went to the node and ran `rm -rf  /usr/local/nvidia/toolkit` and reinstalled the chart. Looks solved to me!

**Comment by elezar** [comment]
@bstrdsmkr that is great news! We will include the centos7 images in our upcoming release. We have something on our backlog to address the mounts not being cleaned up properly, but at least we have a workaround.

**Comment by shivamerla** [comment]
closing this as toolkit version `1.6.0-centos7` is released now, also driver image `470.57.02-centos7`
